### DISCUSSION TITLE
use ts-morph instead of ts-patch

### DISCUSSION THREAD

**Discussion by ryoppippi** [samchon/typia#1118] [Ideas]
ts-patch depends on TypeScript version heavily, and installing ts-patch fails in some environment.
ts-morph is typescript analyzing library that doesn't require patch.
I'm investigating the way to move from ts-patch to ts-morph. I just wanna let you know.

One of my concerns is that I don't know how to integrate with tsc

**Comment by samchon** [comment]
`ts-morph` is TypeScript compiler API wrapper. I know that it can't replace the `ts-patch`.

**Comment by ryoppippi** [comment]
@samchon What are the missing points of TypeScript compiler ? JSDoc?
`ts-morph` supports JSDoc extraction

This is just a discussion

**Reply by samchon** [reply]
The compiler API means features provided by direct importing from TypeScript.

```
import ts from "typescript{";
ts.createProgram(...);
```

**Reply by ryoppippi** [reply]
I know what `TypeScript compiler` means!!
So, what are the missing features and `ts-patch` has?

**Reply by samchon** [reply]
Also, current issue is not of `ts-patch`, but of `typescript`. 

There had not been any problem even in the `typescript@5.5.1-rc` version.

Just break change has happned in the compiler source code level in `typescript@5.5.2`.

**Reply by ryoppippi** [reply]
OK got it, so `ts-patch` needs to be fixed

**Reply by samchon** [reply]
`typescript` supports transformation only in the API level, but not supporting from the `tsc` command. `ts-patch` is a library that ejecting the transformation feature to the `tsc` by hacking the `typescript` in the source code level.

`typescript` team said that they will support the transformation feature through the `tsc` command. However, a lot of time has passed since then, and no one knows whether it is currently in progress or when it will be completed. So that cannot sure when. 

Therefore, `ts-patch` is sole solution for performing transformation through the `tsc` command at now. There was once a library called `ttypescript`, but it was discontinued and disappeared -> https://github.com/cevek/ttypescript

**Reply by samchon** [reply]
Yes, `ts-patch` should be fixed.

**Reply by samchon** [reply]
So I'm waiting this PR to be accepted and newer version to be published -> https://github.com/nonara/ts-patch/pull/161

**Reply by ryoppippi** [reply]
I understand the need for `ts-patch` when using the `tsc` command. However, I have concerns about the significant side effects of `ts-patch`.

While working on a monorepo project, I noticed that it can quickly become unstable under certain conditions. Additionally, when experimenting with `deno-esbuild`, I confirmed that it fails to execute because it doesn't modify the TypeScript core itself.

In my opinion, it might be better to implement this using only the TypeScript Compiler API or `ts-morph` to minimize side effects as much as possible.

While there would be a side effect of not being able to call it directly from `tsc`, I believe providing a dedicated command as an alternative approach could be a good solution. Furthermore, we already have the `generation` command in our CLI tools, and we've also covered major bundlers with `unplugin-typia`.

Of course, this is your OSS project, so the decision ultimately rests with you. I'm simply offering my perspective for your consideration.

P.S.
For the reference, here is the deno code I played

<details><summary>Codes</summary>
<p>

build.ts
```ts
import * as esbuild from "npm:esbuild@0.20.2";
import { denoPlugins } from "jsr:@luca/esbuild-deno-loader@^0.10.3";
import UnpluginTypia from "jsr:@ryoppippi/unplugin-typia/esbuild";

const result = await esbuild.build({
  plugins: [
    UnpluginTypia({/* options */}),
    ...denoPlugins(),
  ],
  entryPoints: ["./index.ts"],
  outfile: "./dist/index.js",
  bundle: true,
  format: "esm",
});

console.log(result.outputFiles);

esbuild.stop();
```

index.ts
```ts
import typia, { type tags } from "npm:typia";

interface IMember {
  email: string & tags.Format<"email">;
  age: number & tags.ExclusiveMinimum<19> & tags.Maximum<100>;
}

const member = {
  email: "example@example.com",
  age: 25,
} as const satisfies IMember;

console.log({ member, is: typia.is<IMember>(member) });

```

tsconfig.json
```json

{
	"compilerOptions": {
		"target": "ESNext",
		"lib": ["esnext", "esnext.asynciterable", "DOM"],
		"module": "NodeNext",
		"moduleResolution": "NodeNext",
		"resolveJsonModule": true,
		"strict": true,
		"strictNullChecks": true,
		"esModuleInterop": true,
		"skipLibCheck": true
	},
	"exclude": ["dist", "eslint.config.js"]
}
```

</p>
</details>

**Reply by samchon** [reply]
It is possible to call `transform()` function of `typia` through the `unplugin-typia` directly instead of using `ts-patch`.

Also, you can take another option of `ts-patch` that not hacking `tsc` but wrapping it through `tspc` command. Of course, you still can replace the `tsc` to the `tspc` in the plugin level.

https://github.com/nonara/ts-patch?tab=readme-ov-file#method-1-live-compiler

**Reply by samchon** [reply]
In `typia` case, as it has started to support NestJS backend server (current [`nestia`](https://github.com/samchon/nestia) project), `typia` has to stick to current position using `tsc` without plugin configuration.

However, `unplugin-typia` is different with `typia`. As `typia` had not supported frontend side, and just proceeded only for backend projects, it is okay `unplugin-typia` to find another setup way with `typia`. It is on your wish, and hope to find better way than me.

**Reply by oscar-b** [reply]
@samchon Does this mean that it will never be possible to use `typia` with `tsx` (https://tsx.is/)? Using `tsx` currently seems to be the recommended way for testing Typescript files in the native Node test runner (https://github.com/nodejs/help/issues/3902).

**Reply by samchon** [reply]
> @samchon Does this mean that it will never be possible to use `typia` with `tsx` (https://tsx.is/)? Using `tsx` currently seems to be the recommended way for testing Typescript files in the native Node test runner ([nodejs/help#3902](https://github.com/nodejs/help/issues/3902)).

@oscar-b Tested `tsx`, but it does not understand `compilerOptions.plugins`.

**Reply by oscar-b** [reply]
@samchon 

> TypeScript & ESM transformations are handled by [esbuild](https://esbuild.github.io/), so it shares some of the same limitations such as:
> Only [certain tsconfig.json properties](https://esbuild.github.io/content-types/#tsconfig-json) are supported
https://esbuild.github.io/content-types/#tsconfig-json

Do you have any other ideas on how to run the node test runner with Typia?

**Reply by samchon** [reply]
@oscar-b I'm doing like this

https://github.com/samchon/nestia-start/blob/master/test/index.ts

https://nestia.io/docs/e2e/development/

**Reply by oscar-b** [reply]
@samchon Thanks, I ended up with something like this for the time being, which seems to work fine:

```
"test": "tsc --watch --project test/tsconfig.json & node --import ./test/out/test/setup.js --watch --test './test/out/src/**/*.test.js'"
```
